emails: Add a custom header specifying the organization when possible.

mateuszmandera opened the pull request:
Closes #15135.

Not sure if this is the right place to put addition of the header, but this way it'll cover all emails where it is viable, without needing to plumb through some kind of `set_organization_header` argument through multiple functions and worrying in setting in all the right email-sending codepaths.

Tested manually in production, by inspecting the headers of the email my server sent to me.
```
...
Date: Sun, 14 Jun 2020 12:08:41 -0000
Message-ID: <159213652145.6636.6569303924278563511@instance-1.us-central1-a.c.sonic-totem-257122.internal>
X-Zulip-Organization: myrealm1
....
```

timabbott commented:
The `X-Zulip-Organization` should probably have the whole `realm.host`, not just the string_id; that's relevant if you're using multiple Zulip servers hosted by different people (they might all have `zulip` as their string id).

timabbott commented on zerver/lib/send_email.py:
```diff
@@ -63,6 +63,10 @@ def build_email(template_prefix: str, to_user_ids: Optional[List[int]]=None,
 
+    extra_headers = {}
+    if len(set([to_user.realm_id for to_user in to_users])) == 1:
+        extra_headers['X-Zulip-Organization'] = to_users[0].realm.string_id
```
I don't like the idea of this auto-detection logic; do we have any emails that we build where `to_users` will be in multiple realms?  I think the answer should be no (since I believe we'll call this in a loop if we're e.g. sending an email to many users across multiple organizations); maybe we should just make it an assert?  

Or perhaps better, maybe we should make `realm` be an argument passed into these functions from the caller?  That would allow e.g. Zulip invitation emails to set this header (well, except for the rare corner case of realm-creation emails).

(Also, we should make at least one email test verify the `X-Zulip-Organization` header).

timabbott commented:
@mateuszmandera ping on this one.

mateuszmandera commented:
I added a realm argument. We have two ways that send_email codepath is called:
1. With `to_user_ids`. In this case we assert they're all from the same realm and set the header based on that.
2. With `to_emails`. In that case we use the `realm` argument passed to the function for the header if possible.

I added verification of the header to various tests for the systems that cause emails to be sent out. But I don't know if it's important to test every single email-sending codepath for header, given that the addition of the header happens at the exit point, that should be shared by them all. I'm not sure if I shouldn't even remove some of those asserts/tests I added and simply have two tests, one for the `to_user_ids` case and one for `to_emails` with `realm` argument?

timabbott commented:
Great!  See https://chat.zulip.org/#narrow/stream/3-backend/topic/X-Zulip-Organization/near/913725 for changes in what we want to encode.

Re: header, I think assert statements in every code path is a great way to ensure this is correct.  It's only a line of code plus a blank line in each email.

mateuszmandera commented:
Switched to `List-Id` header.

> Re: header, I think assert statements in every code path is a great way to ensure this is correct. It's only a line of code plus a blank line in each email.

I'm not sure if you mean asserts in tests or actual `assert` about the sent message in the production code (that wouldn't be possible I think as the codepath just call `send_email` or `queue_json_public('email_senders', ...)` so they can't really assert anything about what is sent out).

timabbott commented:
I just meant the asserts you had already added in the tests.

timabbott commented:
This is great, merged, thanks @mateuszmandera!
